### ISSUE TITLE
NGINX + url rewriting = error 404 (NGINX bug?)

### ISSUE THREAD

**Issue by marioscube** [seblucas/cops#353]
I use COPS + Apache with my kobo e-reader without any problems.

I thought I'd try COPS on NGINX to see if it's any faster on my hardware.
It works like a charm except with $config['cops_use_url_rewriting'] = '1';
And I added:
    	location /download/ {
        	rewrite ^/download/(\d+)/(\d+)/.*\.(.*)$ /fetch.php?data=$1&db=$2&type=$3 last;
        	rewrite ^/download/(\d+)/.*\.(.*)$ /fetch.php?data=$1&type=$2 last;
        	break;
        }
to a working NGINX config file (and rebooting etc.) but I still get the 404 error.

What am I doing wrong?
What do I have to add to COPS config files and or NGINX config files to make url_rewriting work? 

Edit: 
NGINX version 1.6.2
PHP 5.6.30
OS Debian Jessie 8.8 on VirtualBox / WIndows 7 - 32 bits

**Comment by Shadow4711** [comment]
Did you try this [this](https://github.com/seblucas/cops/wiki/Full-example-with-Nginx)?

**Comment by marioscube** [comment]
All, except: 

location /Calibre {
        root /home/calibre/Dropbox;
        internal;
    } 

and

$config['calibre_internal_directory'] = '/Calibre/'; 

I don't get how that should work with my libraries (for testing purposes) in /var/www/html/cops/

Also I did not include in config_local.php
$config['cops_x_accel_redirect'] = "X-Accel-Redirect";
when I did it made no difference.

Does it work for you?

**Comment by Shadow4711** [comment]
I am using Apache and I also have no Kobo reader. Just saw that your posted config is slightly different to the wiki.

**Comment by marioscube** [comment]
I access cops on nginx with http://<lan-ip>/cops/index.php

The rewrite rules (to me) look the same as used on apache where they work.

**Comment by seblucas** [comment]
> I access cops on nginx with http:///cops/index.php

I don't understand, in your url cops is the server name or a directory ?

**Comment by marioscube** [comment]
Problem with accessing github from an iPad.

Http://localhost/cops/index.php

**Comment by Shadow4711** [comment]
Could you try

`rewrite ^/download/(\d+)/(\d+)/.*\.(.*)$ /cops/fetch.php?data=$1&db=$2&type=$3 last;
rewrite ^/download/(\d+)/.*\.(.*)$ /cops/fetch.php?data=$1&type=$2 last;`

**Comment by marioscube** [comment]
Even with several permutations of /cops/ and /download/ i still get the error "404 Not Found" on 2 different virtual machines.

**Comment by Shadow4711** [comment]
Could you please post the relevant part of log?

**Comment by marioscube** [comment]
Something like this?

`2017/05/25 23:14:45 [error] 1632#0: *7 open() "/var/www/html/cops/download/3/0/Poema - Merete Junker.epub" failed (2: No such file or directory), client: 192.168.1.109, server: _, request: "GET /cops/download/3/0/Poema%20-%20Merete%20Junker.epub HTTP/1.1", host: "192.168.1.116", referrer: "http://192.168.1.116/cops/index.php?page=10&db=0"`

**Comment by Shadow4711** [comment]
Seems that the rule is not matching. Did you also change location?

`location /cops/download/ {`
`rewrite ^/cops/download/(\d+)/(\d+)/.*\.(.*)$ /cops/fetch.php?data=$1&db=$2&type=$3 last;`
`rewrite ^/cops/download/(\d+)/.*\.(.*)$ /cops/fetch.php?data=$1&type=$2 last;`
`break;`
`}`

**Comment by marioscube** [comment]
I thought I had tried this permutation as well, but ............

 **this seems to work!!** (I have not yet tried it with a Kobo) ->  Thank you!

However I see no reason why the original rewrite rules do not work.

For me it's no problem, but for some other users it might be a bit harder to understand which nginx config file to edit and why as not all nginx installs are the same (I found out).

But again thank you for COPS!   :D
